cep47-logic/src/lib.rs

```diff
@@ -178,46 +187,25 @@ pub trait CEP47Contract<Storage: CEP47Storage>: WithStorage<Storage> {
         if self.is_paused() {
             return Err(Error::PermissionDenied);
         }
-        let mut sender_tokens = self.storage().get_tokens(sender);
 
         for token_id in token_ids.iter() {
-            if !sender_tokens.contains(token_id) {
-                return Err(Error::PermissionDenied);
+            let token_owner = self.storage().owner_of(token_id);
+            if let Some(token_owner) = token_owner {
+                if &token_owner != sender {
+                    return Err(Error::NotAnOwner);
+                }
             }
```

Comment:
Maybe it's worth extracting to a method - it's used in at least two places already.